metagraph/src/annotation/binary_matrix/row_diff/row_diff.cpp

```diff
@@ -54,34 +75,36 @@ IRowDiff::get_rd_ids(const std::vector<BinaryMatrix::Row> &row_ids) const {
     // been reached before, and thus, will be reconstructed before this one.
     std::vector<std::vector<size_t>> rd_paths_trunc(row_ids.size());
 
-    const graph::boss::BOSS &boss = graph_->get_boss();
-    const bit_vector &rd_succ = fork_succ_.size() ? fork_succ_ : boss.get_last();
-
     for (size_t i = 0; i < row_ids.size(); ++i) {
         Row row = row_ids[i];
 
-        graph::boss::BOSS::edge_index boss_edge = 
-                graph::AnnotatedSequenceGraph::anno_to_graph_index(row);
+        node_index node = graph::AnnotatedSequenceGraph::anno_to_graph_index(row);
 
         while (true) {
-            if (graph_->is_valid(boss_edge)) {
-                row = graph::AnnotatedSequenceGraph::graph_to_anno_index(boss_edge);
-
-                auto [it, is_new] = node_to_rd.try_emplace(row, node_to_rd.size());
-                rd_paths_trunc[i].push_back(it.value());
-
-                // If a node had been reached before, we interrupt the diff path.
-                // The annotation for that node will have been reconstructed earlier
-                // than for other nodes in this path as well. Thus, we will start
-                // reconstruction from that node and don't need its successors.
-                if (!is_new)
-                    break;
-
-                if (anchor_[row])
-                    break;
+            assert(graph_->in_graph(node));
+            row = graph::AnnotatedSequenceGraph::graph_to_anno_index(node);
+
+            auto [it, is_new] = node_to_rd.try_emplace(row, node_to_rd.size());
+            rd_paths_trunc[i].push_back(it.value());
+
+            // If a node had been reached before, we interrupt the diff path.
+            // The annotation for that node will have been reconstructed earlier
+            // than for other nodes in this path as well. Thus, we will start
+            // reconstruction from that node and don't need its successors.
+            if (!is_new)
+                break;
+
+            if (anchor_[row])
+                break;
+
+            if (fork_succ_.size()) {
+                node = row_diff_successor(*graph_, node, fork_succ_);
+            } else { // Only happens in DBGSuccinct
+                auto *dbg_succ = dynamic_cast<graph::DBGSuccinct const*>(graph_);
+                assert(dbg_succ);
+                node = row_diff_successor(*graph_, node, dbg_succ->get_boss().get_last());
```

Comment:
Would it be problematic to, instead of doing this, assign `get_last()` to `fork_succ` during its load, if it was detected that it is empty? I assume in this case we would also need to change `fork_succ_` itself into a pointer to `bit_vector`?